src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/ICustomMarshalerTests.cs

```diff
@@ -0,0 +1,537 @@
+        public static extern int MarshalerOnBoxedValueTypeMethod([MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(BoxedValueTypeCustomMarshaler))] object i);
+
+
+        public class StringContainer
+        {
+            public string Value { get; set; }
+        }
+
+        public class ClassForwardingCustomMarshaler : ICustomMarshaler
+        {
+            public void CleanUpManagedData(object ManagedObj) {}
+            public void CleanUpNativeData(IntPtr pNativeData)
+            {
+                Marshal.ZeroFreeGlobalAllocAnsi(pNativeData);
+            }
+
+            public int GetNativeDataSize() => -1;
+
+            public IntPtr MarshalManagedToNative(object ManagedObj)
+            {
+                return Marshal.StringToCoTaskMemAnsi(((StringContainer)ManagedObj).Value);
+            }
+
+            public object MarshalNativeToManaged(IntPtr pNativeData)
+            {
+                int doubleValue = pNativeData.ToInt32() * 2;
+                return new StringContainer { Value = doubleValue.ToString() };
+            }
+
+            public static ICustomMarshaler GetInstance(string cookie) => new ClassForwardingCustomMarshaler();
+        }
+
+        [DllImport(LibcLibrary, EntryPoint = "atoi", CallingConvention = CallingConvention.Cdecl)]
+        [return: MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(ClassForwardingCustomMarshaler))]
+        public static extern StringContainer MarshalerOnClassTypeMethod([MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(ClassForwardingCustomMarshaler))] StringContainer str);
+
+        public class OrderTrackingCustomMarshaler : ICustomMarshaler
+        {
+            public static List<string> Events { get; } = new List<string>();
+            public static IntPtr MarshaledNativeData { get; set; }
+
+            public void CleanUpManagedData(object ManagedObj)
+            {
+                Events.Add("Called CleanUpManagedData");
+            }
+
+            public void CleanUpNativeData(IntPtr pNativeData)
+            {
+                Assert.Equal(MarshaledNativeData, pNativeData);
+                Marshal.ZeroFreeGlobalAllocAnsi(pNativeData);
+
+                Events.Add("Called CleanUpNativeData");
+            }
+
+            public int GetNativeDataSize()
+            {
+                Events.Add("Called GetNativeDataSize");
+                return 0;
+            }
+
+            public IntPtr MarshalManagedToNative(object ManagedObj)
+            {
+                Events.Add("Called MarshalManagedToNative");
+                MarshaledNativeData = Marshal.StringToCoTaskMemAnsi((string)ManagedObj);
+                return MarshaledNativeData;
+            }
+
+            public object MarshalNativeToManaged(IntPtr pNativeData)
+            {
+                Events.Add("Called MarshalNativeToManaged");
+                return pNativeData.ToInt32().ToString();
```

Comment:
`atoi` returns an `int`.

Since this class is applied to the return parameter of `atoi`, this modifies the `int` value returned from `atoi`, and converts it to `string`. This is similar to the other example of `pNativeData.ToInt32()`

For example, if `atoi` returns `123`, then `pNativeData.ToInt32() == 123`. If I use `Marshal.ReadAsInt32(pNativeData)`, I get an AccessViolationException!